### PR TITLE
Fix template inductive applied to SProp type

### LINKED ISSUE #16784: Assertion failure with SProp
#### Description of the problem

The following code produces an assertion failure:

```coq
Inductive foo (F : Type -> Type) : SProp := Foo (X: Type) (x: F X).

Definition bar (F : Type -> Type) : Type := list (foo F).
```
Result:

```
Anomaly
""Assert_failure _vendor+v8.16+32bit/coq/kernel/inductive.ml:129:19"."
Please report at http://coq.inria.fr/bugs/.
```

It seems like it maybe has something to do with template polymorphism.

#### Coq Version

cow.vercel.app reports v8.16.

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #16784

### CODE REVIEW

File: kernel/inductive.ml [hunk 1]
```diff
@@ -126,7 +126,7 @@ let max_template_universe u v = u @ v
 (* to [x]. *)
 let cons_subst u su subst =
   let su = match su with
-  | Sorts.SProp -> assert false (* No template on SProp *)
+  | Sorts.SProp -> [] (* pretend it's Prop, will fail later *)
```
Comment (ppedrot) [inline kernel/inductive.ml:129]: I think this is too dangerous a fix. You should prevent this from happening somewhere upstream instead.

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by ppedrot** [comment]
@SkySkimmer what about implementing the solution we discussed privately, i.e. having a non-kernel function to typecheck template inductive types in the upper layers?

**Comment by SkySkimmer** [comment]
Because I think the kernel should also produce the proper error rather than assert false

**Comment by SkySkimmer** [comment]
Changed with a full blown solution with proper errors except in retyping (may be changed once we remove implicit cumulative sprop)

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot merge now

**Pull request merged**